Overriding dep doesn't work with peer deps if there is version mismatch

### Verify latest release

- [x] I verified that the issue exists in the latest pnpm release

### pnpm version

10.15.0

### Which area(s) of pnpm are affected? (leave empty if unsure)

Dependencies resolver

### Link to the code that reproduces this issue or a replay of the bug

https://github.com/brc-dd/pnpm-peer-dep-override-bug

### Reproduction steps

Clone the repo and run `pnpm i`.

### Describe the Bug

```
❯ pnpm i
Scope: all 2 workspace projects
/Users/divyansh/foo/foo:
 ERR_PNPM_NO_MATCHING_VERSION  No matching version found for vite@7.1.5 while fetching it from https://registry.npmjs.org/

This error happened while installing a direct dependency of /Users/divyansh/foo/foo

The latest release of vite is "7.1.3".

Other releases are:
  * alpha: 6.0.0-alpha.24
  * previous: 4.5.14
  * beta: 7.1.0-beta.1

If you need the full list of all 668 published versions run "$ pnpm view vite versions".
Progress: resolved 67, reused 45, downloaded 0, added 0
```

In this case `rolldown-vite` has v7.1.5 but `vite` has latest v7.1.3. `foo/package.json` has `@vitejs/plugin-vue` which has a peer dependency on `vite`. pnpm tries to install `vite@7.1.5` instead of respecting override and installing `vite@npm:rolldown-vite@7.1.5`.

### Expected Behavior

Install works fine. It works fine on v10.14.0. I think #9835 broke this?

### Which Node.js version are you using?

24.3.0

### Which operating systems have you used?

- [x] macOS
- [ ] Windows
- [ ] Linux

### If your OS is a Linux based, which one it is? (Include the version if relevant)

_No response_

--- a/.changeset/petite-badgers-rest.md
+++ b/.changeset/petite-badgers-rest.md
@@ -1,0 +1,6 @@
+---
+"@pnpm/resolve-dependencies": patch
+"pnpm": patch
+---
+
+When resolving peer dependencies, pnpm looks whether the peer dependency is present in the root workspace project's dependencies. This change makes it so that the peer dependency is correctly resolved even from aliased npm-hosted dependencies or other types of dependencies [#9913](https://github.com/pnpm/pnpm/issues/9913).

--- a/pkg-manager/core/test/install/peerDependencies.ts
+++ b/pkg-manager/core/test/install/peerDependencies.ts
@@ -1,5 +1,6 @@
 import fs from 'fs'
 import path from 'path'
+import { type Project } from '@pnpm/assert-project'
 import { WANTED_LOCKFILE } from '@pnpm/constants'
 import { type LockfileFile } from '@pnpm/lockfile.fs'
 import { prepareEmpty, preparePackages } from '@pnpm/prepare'
@@ -331,162 +332,122 @@ test('peer dependency is resolved from the dependencies of the workspace root pr
   }
 })
 
-test('peer dependency is resolved from the dependencies of the workspace root project even if there are other versions of the peer dependency present in the dependency graph', async () => {
-  const projects = preparePackages([
-    {
-      location: '.',
-      package: { name: 'root' },
-    },
-    {
-      location: 'pkg',
-      package: {},
-    },
-    {
-      location: 'pkg2',
-      package: {},
-    },
-  ])
-  const allProjects: ProjectOptions[] = [
-    {
-      buildIndex: 0,
-      manifest: {
-        name: 'root',
-        version: '1.0.0',
-
-        dependencies: {
-          ajv: '4.10.0',
-        },
+describe('peer dependency is resolved from the root of the workspace even if there are other versions of the peer dependency present in the dependency graph', () => {
+  let nonRootProjects: ProjectOptions[]
+  let projects: Record<string, Project>
+  let mutatedProjects: MutatedProject[]
+  beforeEach(() => {
+    projects = preparePackages([
+      {
+        location: '.',
+        package: { name: 'root' },
       },
-      rootDir: process.cwd() as ProjectRootDir,
-    },
-    {
-      buildIndex: 0,
-      manifest: {
-        name: 'pkg',
-        version: '1.0.0',
-
-        dependencies: {
-          'ajv-keywords': '1.5.0',
-        },
+      {
+        location: 'pkg',
+        package: {},
       },
-      rootDir: path.resolve('pkg') as ProjectRootDir,
-    },
-    {
-      buildIndex: 0,
-      manifest: {
-        name: 'pkg2',
-        version: '1.0.0',
-
-        dependencies: {
-          ajv: '5.0.0',
-        },
+      {
+        location: 'pkg2',
+        package: {},
       },
-      rootDir: path.resolve('pkg2') as ProjectRootDir,
-    },
-  ]
-  const reporter = jest.fn()
-  await mutateModules([
-    {
-      mutation: 'install',
-      rootDir: process.cwd() as ProjectRootDir,
-    },
-    {
-      mutation: 'install',
-      rootDir: path.resolve('pkg') as ProjectRootDir,
-    },
-    {
-      mutation: 'install',
-      rootDir: path.resolve('pkg2') as ProjectRootDir,
-    },
-  ], testDefaults({ allProjects, reporter, resolvePeersFromWorkspaceRoot: true }))
+    ])
+    nonRootProjects = [
+      {
+        buildIndex: 0,
+        manifest: {
+          name: 'pkg',
+          version: '1.0.0',
 
-  expect(reporter).not.toHaveBeenCalledWith(expect.objectContaining({
-    name: 'pnpm:peer-dependency-issues',
-  }))
-
-  {
-    const lockfile = projects.root.readLockfile()
-    expect(lockfile.importers.pkg?.dependencies?.['ajv-keywords'].version).toBe('1.5.0(ajv@4.10.0)')
-  }
-})
-
-test('peer dependency is resolved from an aliased dependency of the workspace root project even if there are other versions of the peer dependency present in the dependency graph', async () => {
-  const projects = preparePackages([
-    {
-      location: '.',
-      package: { name: 'root' },
-    },
-    {
-      location: 'pkg',
-      package: {},
-    },
-    {
-      location: 'pkg2',
-      package: {},
-    },
-  ])
-  const allProjects: ProjectOptions[] = [
-    {
-      buildIndex: 0,
-      manifest: {
-        name: 'root',
-        version: '1.0.0',
-
-        dependencies: {
-          ajv: 'npm:@pnpm.e2e/foo@100.0.0',
+          dependencies: {
+            'ajv-keywords': '1.5.0',
+          },
         },
+        rootDir: path.resolve('pkg') as ProjectRootDir,
       },
-      rootDir: process.cwd() as ProjectRootDir,
-    },
-    {
-      buildIndex: 0,
-      manifest: {
-        name: 'pkg',
-        version: '1.0.0',
+      {
+        buildIndex: 0,
+        manifest: {
+          name: 'pkg2',
+          version: '1.0.0',
 
-        dependencies: {
-          'ajv-keywords': '1.5.0',
+          dependencies: {
+            ajv: '5.0.0',
+          },
         },
+        rootDir: path.resolve('pkg2') as ProjectRootDir,
       },
-      rootDir: path.resolve('pkg') as ProjectRootDir,
-    },
-    {
-      buildIndex: 0,
-      manifest: {
-        name: 'pkg2',
-        version: '1.0.0',
+    ]
+    mutatedProjects = [
+      {
+        mutation: 'install',
+        rootDir: process.cwd() as ProjectRootDir,
+      },
+      {
+        mutation: 'install',
+        rootDir: path.resolve('pkg') as ProjectRootDir,
+      },
+      {
+        mutation: 'install',
+        rootDir: path.resolve('pkg2') as ProjectRootDir,
+      },
+    ]
+  })
+  test('the package in the root is a regular non-aliased npm-hosted dependency', async () => {
+    const allProjects: ProjectOptions[] = [
+      {
+        buildIndex: 0,
+        manifest: {
+          name: 'root',
+          version: '1.0.0',
 
-        dependencies: {
-          ajv: '5.0.0',
+          dependencies: {
+            ajv: '4.10.0',
+          },
         },
+        rootDir: process.cwd() as ProjectRootDir,
       },
-      rootDir: path.resolve('pkg2') as ProjectRootDir,
-    },
-  ]
-  const reporter = jest.fn()
-  await mutateModules([
-    {
-      mutation: 'install',
-      rootDir: process.cwd() as ProjectRootDir,
-    },
-    {
-      mutation: 'install',
-      rootDir: path.resolve('pkg') as ProjectRootDir,
-    },
-    {
-      mutation: 'install',
-      rootDir: path.resolve('pkg2') as ProjectRootDir,
-    },
-  ], testDefaults({ allProjects, reporter, resolvePeersFromWorkspaceRoot: true }))
+      ...nonRootProjects,
+    ]
+    const reporter = jest.fn()
+    await mutateModules(mutatedProjects, testDefaults({ allProjects, reporter, resolvePeersFromWorkspaceRoot: true }))
 
-  expect(reporter).not.toHaveBeenCalledWith(expect.objectContaining({
-    name: 'pnpm:peer-dependency-issues',
-  }))
+    expect(reporter).not.toHaveBeenCalledWith(expect.objectContaining({
+      name: 'pnpm:peer-dependency-issues',
+    }))
 
-  {
-    const lockfile = projects.root.readLockfile()
-    expect(lockfile.importers.pkg?.dependencies?.['ajv-keywords'].version).toBe('1.5.0(@pnpm.e2e/foo@100.0.0)')
-  }
+    {
+      const lockfile = projects.root.readLockfile()
+      expect(lockfile.importers.pkg?.dependencies?.['ajv-keywords'].version).toBe('1.5.0(ajv@4.10.0)')
+    }
+  })
+  test('the package in the root is aliased', async () => {
+    const allProjects: ProjectOptions[] = [
+      {
+        buildIndex: 0,
+        manifest: {
+          name: 'root',
+          version: '1.0.0',
+
+          dependencies: {
+            ajv: 'npm:@pnpm.e2e/foo@100.0.0',
+          },
+        },
+        rootDir: process.cwd() as ProjectRootDir,
+      },
+      ...nonRootProjects,
+    ]
+    const reporter = jest.fn()
+    await mutateModules(mutatedProjects, testDefaults({ allProjects, reporter, resolvePeersFromWorkspaceRoot: true }))
+
+    expect(reporter).not.toHaveBeenCalledWith(expect.objectContaining({
+      name: 'pnpm:peer-dependency-issues',
+    }))
+
+    {
+      const lockfile = projects.root.readLockfile()
+      expect(lockfile.importers.pkg?.dependencies?.['ajv-keywords'].version).toBe('1.5.0(@pnpm.e2e/foo@100.0.0)')
+    }
+  })
 })
 
 test('warning is reported when cannot resolve peer dependency for non-top-level dependency', async () => {

--- a/pkg-manager/core/test/install/peerDependencies.ts
+++ b/pkg-manager/core/test/install/peerDependencies.ts
@@ -420,7 +420,7 @@ describe('peer dependency is resolved from the root of the workspace even if the
       expect(lockfile.importers.pkg?.dependencies?.['ajv-keywords'].version).toBe('1.5.0(ajv@4.10.0)')
     }
   })
-  test('the package in the root is aliased', async () => {
+  test('the package in the root is aliasing a package with a different name', async () => {
     const allProjects: ProjectOptions[] = [
       {
         buildIndex: 0,
@@ -446,6 +446,36 @@ describe('peer dependency is resolved from the root of the workspace even if the
     {
       const lockfile = projects.root.readLockfile()
       expect(lockfile.importers.pkg?.dependencies?.['ajv-keywords'].version).toBe('1.5.0(@pnpm.e2e/foo@100.0.0)')
+    }
+  })
+  test('the package in the root is under an alias', async () => {
+    const allProjects: ProjectOptions[] = [
+      {
+        buildIndex: 0,
+        manifest: {
+          name: 'root',
+          version: '1.0.0',
+
+          dependencies: {
+            b: 'npm:ajv@1.0.0',
+            a: 'npm:ajv@4.10.0',
+            c: 'npm:ajv@2.0.0',
+          },
+        },
+        rootDir: process.cwd() as ProjectRootDir,
+      },
+      ...nonRootProjects,
+    ]
+    const reporter = jest.fn()
+    await mutateModules(mutatedProjects, testDefaults({ allProjects, reporter, resolvePeersFromWorkspaceRoot: true }))
+
+    expect(reporter).not.toHaveBeenCalledWith(expect.objectContaining({
+      name: 'pnpm:peer-dependency-issues',
+    }))
+
+    {
+      const lockfile = projects.root.readLockfile()
+      expect(lockfile.importers.pkg?.dependencies?.['ajv-keywords'].version).toBe('1.5.0(ajv@4.10.0)')
     }
   })
 })

--- a/pkg-manager/core/test/install/peerDependencies.ts
+++ b/pkg-manager/core/test/install/peerDependencies.ts
@@ -410,6 +410,85 @@ test('peer dependency is resolved from the dependencies of the workspace root pr
   }
 })
 
+test('peer dependency is resolved from an aliased dependency of the workspace root project even if there are other versions of the peer dependency present in the dependency graph', async () => {
+  const projects = preparePackages([
+    {
+      location: '.',
+      package: { name: 'root' },
+    },
+    {
+      location: 'pkg',
+      package: {},
+    },
+    {
+      location: 'pkg2',
+      package: {},
+    },
+  ])
+  const allProjects: ProjectOptions[] = [
+    {
+      buildIndex: 0,
+      manifest: {
+        name: 'root',
+        version: '1.0.0',
+
+        dependencies: {
+          ajv: 'npm:@pnpm.e2e/foo@100.0.0',
+        },
+      },
+      rootDir: process.cwd() as ProjectRootDir,
+    },
+    {
+      buildIndex: 0,
+      manifest: {
+        name: 'pkg',
+        version: '1.0.0',
+
+        dependencies: {
+          'ajv-keywords': '1.5.0',
+        },
+      },
+      rootDir: path.resolve('pkg') as ProjectRootDir,
+    },
+    {
+      buildIndex: 0,
+      manifest: {
+        name: 'pkg2',
+        version: '1.0.0',
+
+        dependencies: {
+          ajv: '5.0.0',
+        },
+      },
+      rootDir: path.resolve('pkg2') as ProjectRootDir,
+    },
+  ]
+  const reporter = jest.fn()
+  await mutateModules([
+    {
+      mutation: 'install',
+      rootDir: process.cwd() as ProjectRootDir,
+    },
+    {
+      mutation: 'install',
+      rootDir: path.resolve('pkg') as ProjectRootDir,
+    },
+    {
+      mutation: 'install',
+      rootDir: path.resolve('pkg2') as ProjectRootDir,
+    },
+  ], testDefaults({ allProjects, reporter, resolvePeersFromWorkspaceRoot: true }))
+
+  expect(reporter).not.toHaveBeenCalledWith(expect.objectContaining({
+    name: 'pnpm:peer-dependency-issues',
+  }))
+
+  {
+    const lockfile = projects.root.readLockfile()
+    expect(lockfile.importers.pkg?.dependencies?.['ajv-keywords'].version).toBe('1.5.0(@pnpm.e2e/foo@100.0.0)')
+  }
+})
+
 test('warning is reported when cannot resolve peer dependency for non-top-level dependency', async () => {
   prepareEmpty()
   await addDistTag({ package: '@pnpm.e2e/abc-parent-with-ab', version: '1.0.0', distTag: 'latest' })

--- a/pkg-manager/resolve-dependencies/package.json
+++ b/pkg-manager/resolve-dependencies/package.json
@@ -53,6 +53,7 @@
     "@pnpm/semver.peer-range": "workspace:*",
     "@pnpm/store-controller-types": "workspace:*",
     "@pnpm/types": "workspace:*",
+    "@pnpm/util.lex-comparator": "catalog:",
     "@pnpm/workspace.spec-parser": "workspace:*",
     "@yarnpkg/core": "catalog:",
     "filenamify": "catalog:",

--- a/pkg-manager/resolve-dependencies/src/dedupeInjectedDeps.ts
+++ b/pkg-manager/resolve-dependencies/src/dedupeInjectedDeps.ts
@@ -82,6 +82,7 @@ function applyDedupeMap<T extends PartialResolvedPackage> (
       const prev = opts.resolvedImporters[id].directDependencies[index]
       const linkedDep: LinkedDependency & ResolvedDirectDependency = {
         ...prev,
+        pkg: prev,
         isLinkedDependency: true,
         pkgId: `link:${normalize(path.relative(id, dedupedProjectId))}` as PkgResolutionId,
         resolution: {

--- a/pkg-manager/resolve-dependencies/src/hoistPeers.ts
+++ b/pkg-manager/resolve-dependencies/src/hoistPeers.ts
@@ -12,7 +12,7 @@ export function hoistPeers (
 ): Record<string, string> {
   const dependencies: Record<string, string> = {}
   for (const [peerName, { range }] of missingRequiredPeers) {
-    const rootDep = opts.workspaceRootDeps.find((rootDep) => rootDep.alias === peerName)
+    const rootDep = opts.workspaceRootDeps.find((rootDep) => rootDep.pkg.name === peerName)
     if (rootDep?.version) {
       dependencies[peerName] = rootDep.version
       continue

--- a/pkg-manager/resolve-dependencies/src/hoistPeers.ts
+++ b/pkg-manager/resolve-dependencies/src/hoistPeers.ts
@@ -12,14 +12,14 @@ export function hoistPeers (
 ): Record<string, string> {
   const dependencies: Record<string, string> = {}
   for (const [peerName, { range }] of missingRequiredPeers) {
-    const rootDep = opts.workspaceRootDeps.find((rootDep) => rootDep.pkg.name === peerName)
-    if (rootDep?.version) {
-      dependencies[peerName] = rootDep.version
-      continue
-    }
     const rootDepByAlias = opts.workspaceRootDeps.find((rootDep) => rootDep.alias === peerName)
     if (rootDepByAlias?.normalizedBareSpecifier) {
       dependencies[peerName] = rootDepByAlias.normalizedBareSpecifier
+      continue
+    }
+    const rootDep = opts.workspaceRootDeps.find((rootDep) => rootDep.pkg.name === peerName)
+    if (rootDep?.normalizedBareSpecifier) {
+      dependencies[peerName] = rootDep.normalizedBareSpecifier
       continue
     }
     if (opts.allPreferredVersions![peerName]) {

--- a/pkg-manager/resolve-dependencies/src/hoistPeers.ts
+++ b/pkg-manager/resolve-dependencies/src/hoistPeers.ts
@@ -1,4 +1,5 @@
 import { type PreferredVersions } from '@pnpm/resolver-base'
+import { lexCompare } from '@pnpm/util.lex-comparator'
 import semver from 'semver'
 import { type PkgAddressOrLink } from './resolveDependencies.js'
 
@@ -17,7 +18,9 @@ export function hoistPeers (
       dependencies[peerName] = rootDepByAlias.normalizedBareSpecifier
       continue
     }
-    const rootDep = opts.workspaceRootDeps.find((rootDep) => rootDep.pkg.name === peerName)
+    const rootDep = opts.workspaceRootDeps
+      .filter((rootDep) => rootDep.pkg.name === peerName)
+      .sort((rootDep1, rootDep2) => lexCompare(rootDep1.alias, rootDep2.alias))[0]
     if (rootDep?.normalizedBareSpecifier) {
       dependencies[peerName] = rootDep.normalizedBareSpecifier
       continue

--- a/pkg-manager/resolve-dependencies/src/hoistPeers.ts
+++ b/pkg-manager/resolve-dependencies/src/hoistPeers.ts
@@ -17,6 +17,11 @@ export function hoistPeers (
       dependencies[peerName] = rootDep.version
       continue
     }
+    const rootDepByAlias = opts.workspaceRootDeps.find((rootDep) => rootDep.alias === peerName)
+    if (rootDepByAlias?.normalizedBareSpecifier) {
+      dependencies[peerName] = rootDepByAlias.normalizedBareSpecifier
+      continue
+    }
     if (opts.allPreferredVersions![peerName]) {
       const versions: string[] = []
       const nonVersions: string[] = []

--- a/pkg-manager/resolve-dependencies/src/resolveDependencies.ts
+++ b/pkg-manager/resolve-dependencies/src/resolveDependencies.ts
@@ -102,18 +102,21 @@ DependenciesTreeNode<T>
 
 export type ResolvedPkgsById = Record<PkgResolutionId, ResolvedPackage>
 
-export interface LinkedDependency {
-  isLinkedDependency: true
-  optional: boolean
-  dev: boolean
-  resolution: DirectoryResolution
-  pkgId: PkgResolutionId
-  pkg: PackageManifest
-  version: string
-  name: string
+export interface PkgAddressOrLinkBase {
   alias: string
   catalogLookup?: CatalogLookupMetadata
   normalizedBareSpecifier?: string
+  optional: boolean
+  pkg: PackageManifest
+  pkgId: PkgResolutionId
+}
+
+export interface LinkedDependency extends PkgAddressOrLinkBase {
+  isLinkedDependency: true
+  dev: boolean
+  resolution: DirectoryResolution
+  version: string
+  name: string
 }
 
 export interface PendingNode {
@@ -191,32 +194,21 @@ interface MissingPeersOfChildren {
   resolved?: boolean
 }
 
-export type PkgAddress = {
-  alias: string
+export interface PkgAddress extends PkgAddressOrLinkBase {
   depIsLinked: boolean
   isNew: boolean
   isLinkedDependency?: false
   resolvedVia?: string
   nodeId: NodeId
-  pkgId: PkgResolutionId
   installable: boolean
-  pkg: PackageManifest
   version?: string
   updated: boolean
   rootDir: string
   missingPeers: MissingPeers
   missingPeersOfChildren?: MissingPeersOfChildren
   publishedAt?: string
-  catalogLookup?: CatalogLookupMetadata
-  optional: boolean
-  normalizedBareSpecifier?: string
   saveCatalogName?: string
-} & ({
-  isLinkedDependency: true
-  version: string
-} | {
-  isLinkedDependency: undefined
-})
+}
 
 export type PkgAddressOrLink = PkgAddress | LinkedDependency
 

--- a/pkg-manager/resolve-dependencies/src/resolveDependencies.ts
+++ b/pkg-manager/resolve-dependencies/src/resolveDependencies.ts
@@ -108,6 +108,7 @@ export interface LinkedDependency {
   dev: boolean
   resolution: DirectoryResolution
   pkgId: PkgResolutionId
+  pkg: PackageManifest
   version: string
   name: string
   alias: string
@@ -1406,6 +1407,7 @@ async function resolveDependency (
       resolution: pkgResponse.body.resolution,
       version: pkgResponse.body.manifest.version,
       normalizedBareSpecifier: pkgResponse.body.normalizedBareSpecifier,
+      pkg: pkgResponse.body.manifest,
     }
   }
 

